File: integer-to-roman.java
public class Solution {
    char[] c10 = {'I', 'X', 'C', 'M'};
    char[] c5 = {'V', 'L', 'D'};

    public String intToRoman(int num) {
        StringBuilder ans = new StringBuilder();
        int pos = 0;
        while (num > 0) {
            int digit = num % 10;
            if (digit <= 3) {
                for (int i = 0; i < digit; i++)
                    ans.append(c10[pos]);
            } else if (digit == 4) {
                ans.append(c5[pos]);
                ans.append(c10[pos]);
            } else if (digit <= 8) {
                for (int i = 6; i <= digit; i++)
                    ans.append(c10[pos]);
                ans.append(c5[pos]);
            } else {
                ans.append(c10[pos + 1]);
                if (digit == 9) ans.append(c10[pos]);
            }
            num /= 10;
            pos++;
        }
        ans.reverse();
        return ans.toString();
    }

    public static void main(String args[]) {
        new Solution().intToRoman(3173);
    }
}

File: integer-to-roman.cc
#include <string>
using namespace std;

class Solution {
public:
  string intToRoman(int num) {
    static const char one[] = {'I', 'X', 'C', 'M'};
    static const char five[] = {'V', 'L', 'D'};
    int digit[4];
    for (int i = 0; i < 4; i++) {
      digit[i] = num % 10;
      num /= 10;
    }
    string ans;
    for (int i = 3; i >= 0; i--) {
      if (digit[i] == 0) continue;
      if (digit[i] <= 3) {
        for (int j = 0; j < digit[i]; j++) ans += one[i];
      } else if (digit[i] == 4) {
        ans += one[i];
        ans += five[i];
      } else if (digit[i] <= 8) {
        ans += five[i];
        for (int j = 5; j < digit[i]; j++) ans += one[i];
      } else if (digit[i] == 9) {
        ans += one[i];
        ans += one[i + 1];
      }
    }
    return ans;
  }
};
